src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -1250,32 +1250,74 @@ void ObjectMonitor::vthread_epilog(JavaThread* current, ObjectWaiter* node) {
   }
 }
 
-// Return the tail of the _entry_list. If the tail is currently not
-// known, find it by walking from the head of _entry_list, and while
-// doing so assign the _prev pointers to create a doubly linked list.
-ObjectWaiter* ObjectMonitor::entry_list_tail(JavaThread* current) {
+// Convert entry_list into a doubly linked list by assigning the prev
+// pointers and the entry_list_tail pointer. Within the entry_list the
+// next pointers always form a consistent singly linked list. The
+// entry_list can be in either of these forms:
+//   1. Empty.
+//   2. Only singly linked.
+//   3. Only doubly linked.
+//   4. Starting as singly linked (from the head), but ending as doubly
+//      linked (at the tail).
+// Number four is because new threads has pushed themself to the
+// entry_list head after the entry_list was last converted into a
+// doubly linked list.
+void ObjectMonitor::entry_list_build_dll(JavaThread* current) {
   assert(has_owner(current), "invariant");
-  ObjectWaiter* w = _entry_list_tail;
-  if (w != nullptr) {
-    return w;
-  }
-  // Need acquire here to match the implicit release of the cmpxchg
-  // that updated _entry_list, so we can access w->_next.
-  w = Atomic::load_acquire(&_entry_list);
-  assert(w != nullptr, "invariant");
-  if (w->next() == nullptr) {
-    _entry_list_tail = w;
-    return w;
-  }
   ObjectWaiter* prev = nullptr;
+  // Need acquire here to match the implicit release of the cmpxchg
+  // that updated entry_list, so we can access w->prev().
+  ObjectWaiter* w = Atomic::load_acquire(&_entry_list);
   while (w != nullptr) {
     assert(w->TState == ObjectWaiter::TS_ENTER, "invariant");
+    assert(w->prev() == nullptr || w->prev() == prev, "should be");
+    if (w->prev() != nullptr) {
+      break;
+    }
     w->_prev = prev;
     prev = w;
     w = w->next();
   }
-  _entry_list_tail = prev;
-  return prev;
+  if (w == nullptr) {
+    // We converted the entire entry_list from a singly linked list
+    // into a doubly linked list. Now we just need to set the tail
+    // pointer.
```

Comment:
This could also be the empty list case so comment could be improved to separate those two cases. No need to change it if we assert list can’t be empty at the beginning.